### PR TITLE
Add some mod options to AcceptsDeliveredCash and DeliversCash

### PR DESCRIPTION
**Pull request by reaperrr**
Supersedes #18109.

Rebased and applied the following fixes/adjustments:
- renamed `PercentagePayload` to `ValuePercentagePayload`
- adjusted descriptions to make clear the above and `Payload` are additive
- added update rule (tested)
- added some misc fixes for issues I came across while writing the update rule

Nothing major since #18109 was reviewed by @Mailaender, so I'm adding one +1 as starting point.

### CODE REVIEW

**Comment by pchote** [comment]
I'm sorry to have to point out again that the changes in this PR look fine, but reveal underlying bogus behaviour in the existing code. The target owner checks are only applied in the order targeter and not validated in the `DonateCash` activity (and `DonateExperience` has the same problem). This means that cash/experience can be delivered to the wrong player if the target is captured while the unit is moving, or if the player deliberately modifies their client.

This should be fixed by adding a `TryStartEnter` override that checks `ValidRelationships` and `SamePlayerOnly`, and again in `OnEnterComplete`.

**Comment by reaperrr** [comment]
Updated accordingly (untested, though).

**Comment by Smittytron** [comment]
Needs rebase.

**Comment by reaperrr** [comment]
Rebased.

File: OpenRA.Mods.Common/Activities/DonateCash.cs [hunk 1]
```diff
@@ -21,16 +21,52 @@ class DonateCash : Enter
...
+		{
+			enterActor = targetActor;
+
+			// Make sure the target actor is still owned by a player with eligible relationship
+			var isTargetStillValid = true;
+			var targetInfo = targetActor.Info.TraitInfoOrDefault<AcceptsDeliveredCashInfo>();
+			if (targetInfo == null
+				|| !targetInfo.ValidRelationships.HasRelationship(targetActor.Owner.RelationshipWith(self.Owner))
+				|| (targetInfo.SamePlayerOnly && targetActor.Owner != self.Owner))
+				isTargetStillValid = false;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Activities/DonateCash.cs:43]: Could we extract this to a new method called `ValidTarget` or something and use it here and below?

File: OpenRA.Mods.Common/Activities/DonateExperience.cs [hunk 2]
```diff
@@ -35,7 +35,13 @@ protected override bool TryStartEnter(Actor self, Actor targetActor)
 			enterActor = targetActor;
 			enterGainsExperience = targetActor.TraitOrDefault<GainsExperience>();
 
-			if (enterGainsExperience == null || enterGainsExperience.Level == enterGainsExperience.MaxLevel)
+			// Make sure the target actor is still owned by a player with eligible relationship
+			var isTargetStillValid = true;
+			var targetInfo = targetActor.Info.TraitInfoOrDefault<AcceptsDeliveredExperienceInfo>();
+			if (targetInfo == null || !targetInfo.ValidRelationships.HasRelationship(targetActor.Owner.RelationshipWith(self.Owner)))
+				isTargetStillValid = false;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Activities/DonateExperience.cs:42]: Same here

**Comment by PunkPun** [comment]
Closing as cleanup. This code is fairly outdated and will become even more so with the docking refactor and the upcoming resource refactor in https://discord.com/channels/153649279762694144/1023312486600024137

Linking this to #18087 and adding to the future milestone so we don't forget about this

**Pull request closed without merging**